Fix file.managed for Windows

twangboy opened the pull request:
### What does this PR do?

Fixes file.managed for windows
### What issues does this PR fix or reference?

https://github.com/saltstack/zh/issues/697
For some reason, there is a problem with the following code when run from the file.py module:

```
from salt.ext import six
comment = 'This is a string'
isinstance(comment, six.string_types)
```

When run from within the python shell it works fine. Had to revert back to the old import:

```
from salt.ext.six import string_types
```
### Previous Behavior

file.managed would fail with the following error:

```
Unable to manage file: global name 'six' is not defined
```
### New Behavior

`file.managed` works as expected
### Tests written?

No

twangboy committed:
Revert back to import string_types

For some reason, there is a problem with the following
code when run from the file.py module:
```
from salt.ext import six
comment = 'This is a string'
isinstance(comment, six.string_types)
```
When run from within the python shell it works fine.

twangboy committed:
Add six import

twangboy commented:
This will need to be merged forward to all branches

twangboy committed:
Fix some lint

twangboy commented:
DO NOT MERGE THIS FORWARD PLEASE

skizunov commented:
I am curious, did you try:

```
import salt.ext.six as six
```

Which is what is used in most places, instead of what you had:

```
from salt.ext import six
```

twangboy commented:
@skizunov I will try that... thanks!

twangboy committed:
Use correct six import

skizunov commented on salt/modules/file.py:
```diff
@@ -33,7 +33,9 @@
 # pylint: disable=import-error,no-name-in-module,redefined-builtin
-from salt.ext import six
+# This string_types import needs to be here for windows as six.string_types
+# does not work from this module...
```
This comment no longer applies.

twangboy committed:
Changed it back to explicit import

twangboy commented:
@skizunov For some reason `import salt.ext.six as six` doesn't work either. This is only effecting the 2015.5 branch. The code works beautifully in later branches. I did a git bisect and found this PR is where it broke: https://github.com/saltstack/salt/pull/32691

I've reverted back to the `from salt.ext.six import string_types` as that's the only thing working...

twangboy committed:
Additional comments specific to 2015.5

cachedout commented:
I am not going to merge a PR if we don't know why (or how) it works. This could well be the proper solution, but I want to have a full understanding of what's going on here, otherwise, I'm very concern that this might introduce unexpected side-effects.

@twangboy Could you please provide background on the problem that's actually occurring here?

cachedout commented:
Is this a clean install? I am really wondering if this is stale *.pyc files. I can't see any reason at all why this would make sense.

skizunov commented:
What confuses me is that the title for this PR is "Fix file.managed for Windows" but if you look at the modified file, it contains:

```
def __virtual__():
    '''
    Only work on POSIX-like systems
    '''
    # win_file takes care of windows
    if salt.utils.is_windows():
        return False
    return True
```

So the LazyLoader shouldn't load it on Windows. On Windows, it should be loading `win_file.py` instead.

cachedout commented:
@skizunov Funny, I was *just* about to write that same thing. I am very confused here. This isn't making sense to me.

twangboy commented:
@skizunov @cachedout `win_file.py` imports a bunch of functions from `file.py`, one of which is `check_perms`.

twangboy commented:
@cachedout Here's the stack trace:
```
[DEBUG   ] Traceback (most recent call last):
  File "c:\salt-dev\salt\salt\states\file.py", line 1586, in managed
    follow_symlinks)
  File "c:\salt-dev\salt\salt\modules\file.py", line 3756, in manage_file
    ret, _ = check_perms(name, ret, user, group, mode, follow_symlinks)
  File "c:\salt-dev\salt\salt\modules\file.py", line 3243, in check_perms
    if isinstance(orig_comment, six.string_types):
NameError: global name 'six' is not defined
```

This is a clean install. I did a `git reset --hard` and a `git clean -fxd` on the salt clone. I removed python from the system and removed the `C:\python27` folder before reinstalling. I do this everytime I do a build to make sure it's all clean.

cachedout commented:
Ahhh, that makes slightly more sense. I think I see what's happening here. We're importing it in one way in `win_file` and another here, but we're overloading the global namespace. Still digging to figure out exactly _why_ that's the case, but if so, this fix likely needs to be over on the win_file side of the house, IMHO.

twangboy commented:
@cachedout Oooooooh. We probably need to put the six import in win_file....

cachedout commented:
Yeah, that will probably do it. This is why I *hate* the way we do these modules, but that's a story for another day. :] If that fixes it, please update this PR and then I'll get it in.

twangboy commented:
Awesome

twangboy commented:
That fixes it, I'll change the PR

cachedout commented:
K, please isolate the changes to the win_file side and pull out these and we'll be good. Thanks!

twangboy committed:
Fix file.managed for real

twangboy committed:
Move comment to clarify purpose

twangboy commented:
@cachedout I left in @skizunov 's recommended import method

cachedout commented:
@twangboy I'm cool with that. Thanks.

cachedout commented:
@rallytime This is good to merge after the tests pass.

Thanks to @twangboy and @skizunov for the good work here.
